Race condition on generate

A recent CI/CD test on a documentation change uncovered a race condition:

```sh
=== RUN   TestGenerateJSON
time="2019-04-18T14:26:16Z" level=info msg="💾  Loading ../test/fixtures/generate/component.json"
time="2019-04-18T14:26:16Z" level=info msg="💾  Loading ../test/fixtures/generate/infra/component.json"
time="2019-04-18T14:26:16Z" level=info msg="💾  Loading ../test/fixtures/generate/infra/config/prod.json"
time="2019-04-18T14:26:16Z" level=info msg="💾  Loading ../test/fixtures/generate/infra/config/common.json"
time="2019-04-18T14:26:16Z" level=info msg="💾  Writing ../test/fixtures/generate/generated/prod-east-prod/microservices-workload.yaml"
time="2019-04-18T14:26:16Z" level=info msg="🙌  finished generate"
==================
WARNING: DATA RACE
Read at 0x00c0001ce130 by goroutine 12:
  runtime.chansend()
      /usr/local/go1.11/src/runtime/chan.go:140 +0x0
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func5.1()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:286 +0x1d0

Previous write at 0x00c0001ce130 by goroutine 11:
  runtime.closechan()
      /usr/local/go1.11/src/runtime/chan.go:327 +0x0
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func1()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:246 +0x285
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func5.1.1.1()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:293 +0x139
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func5.1.1()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:299 +0xf4

Goroutine 12 (running) created at:
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func5()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:282 +0x14a

Goroutine 11 (running) created at:
  github.com/Microsoft/fabrikate/core.WalkComponentTree.func5.1()
      /home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:289 +0x20f
==================
panic: send on closed channel

goroutine 21 [running]:
github.com/Microsoft/fabrikate/core.WalkComponentTree.func5.1(0xc0001ce120, 0xbea2f8, 0xc000198480, 0xc000188450, 0xc0001984a0, 0xc00002ac88, 0x5, 0x0, 0x0, 0x0, ...)
	/home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:286 +0x1d1
created by github.com/Microsoft/fabrikate/core.WalkComponentTree.func5
	/home/vsts/work/1/s/gopath/src/github.com/Microsoft/fabrikate/core/component.go:282 +0x14b
FAIL	github.com/Microsoft/fabrikate/cmd	0.026s
```

I am running the tests on CI/CD with the -race flag:

```sh
$ go test -v -race ./...
```

--- a/cmd/add_test.go
+++ b/cmd/add_test.go
@@ -24,7 +24,7 @@ func TestAdd(t *testing.T) {
 
 	componentComponent := core.Component{
 		Name:      "cloud-native",
-		Source:    "https://github.com/timfpark/fabriakte-cloud-native",
+		Source:    "https://github.com/timfpark/fabrikate-cloud-native",
 		Method:    "git",
 		Generator: "component",
 	}

--- a/cmd/install.go
+++ b/cmd/install.go
@@ -13,6 +13,15 @@ import (
 
 // Install installs the component at the given path and all of its subcomponents.
 func Install(path string) (err error) {
+	// Make sure host system contains all utils needed by Fabrikate
+	requiredSystemTools := []string{"git", "helm", "sh", "curl"}
+	for _, tool := range requiredSystemTools {
+		path, err := exec.LookPath(tool)
+		if err != nil {
+			return err
+		}
+		log.Info(emoji.Sprintf(":mag: using %s: %s", tool, path))
+	}
 
 	log.Info(emoji.Sprintf(":point_right: Initializing Helm"))
 	if err = exec.Command("helm", "init", "--client-only").Run(); err != nil {

--- a/core/component.go
+++ b/core/component.go
@@ -115,7 +115,7 @@ func (c *Component) ExecuteHook(hook string) (err error) {
 	for _, command := range c.Hooks[hook] {
 		log.Info(emoji.Sprintf(":fishing_pole_and_fish: executing command: %s", command))
 		if len(command) != 0 {
-			cmd := exec.Command("bash", "-c", command)
+			cmd := exec.Command("sh", "-c", command)
 			cmd.Dir = c.PhysicalPath
 			out, err := cmd.Output()
 
@@ -231,28 +231,10 @@ type WalkResult struct {
 //
 // Same level ordering is not ensured; any nodes on the same tree level can be visited in any order.
 // Parent->Child ordering is ensured; A parent is always visited via `iterator` before the children are visited.
-func WalkComponentTree(startingPath string, environments []string, iterator ComponentIteration) chan WalkResult {
+func WalkComponentTree(startingPath string, environments []string, iterator ComponentIteration) <-chan WalkResult {
 	queue := make(chan Component)                // components enqueued to be 'visited' (ie; walked over)
 	results := make(chan WalkResult)             // To pass WalkResults to
 	workingCounter := make(chan Component, 1024) // A counter of components currently being worked on; used to determine if operation complete. An arbitrary large size is chosen. Must be able to hold the entire tree.
-
-	// Closes the returned channels if their are no more components to be worked on
-	markComponentAsVisited := func(completedComponent Component) {
-		results <- WalkResult{Component: &completedComponent}
-		<-workingCounter
-		log.Debugf(fmt.Sprintf("working counter decremented to %d by %s", len(workingCounter), completedComponent.Name))
-		if len(workingCounter) == 0 && len(queue) == 0 {
-			log.Debugf("closing completed and error channels")
-			close(results)
-		}
-	}
-
-	// Increments the working counter; adding `componentToEnqueue` to the queue
-	enqueueComponent := func(componentToEnqueue Component) {
-		queue <- componentToEnqueue
-		workingCounter <- componentToEnqueue
-		log.Debugf(fmt.Sprintf("working counter incremented to %d by %s", len(workingCounter), componentToEnqueue.Name))
-	}
 
 	// Prepares `component` by loading/de-serializing the component.yaml/json and configs
 	// Note: this is only needed for non-inlined components
@@ -266,39 +248,47 @@ func WalkComponentTree(startingPath string, environments []string, iterator Comp
 		return component
 	}
 
+	// Increments the working counter
+	incrementWorkingCounter := func(component Component) {
+		workingCounter <- component
+		log.Debugf(fmt.Sprintf("working counter incremented to %d by %s", len(workingCounter), component.Name))
+	}
+
+	// Decrements the working counter and sends `component` to results channel.
+	// Closes the result channel if workingCounter reaches 0; meaning all nodes in tree have been walked.
+	decrementWorkingCounter := func(component Component) {
+		results <- WalkResult{Component: &component}
+		<-workingCounter
+		log.Debugf(fmt.Sprintf("working counter decremented to %d by %s", len(workingCounter), component.Name))
+		if len(workingCounter) == 0 {
+			log.Debugf("closing results channel")
+			close(results)
+		}
+	}
+
+	// Enqueue the given component
+	enqueue := func(component Component) {
+		// Increment working counter; MUST happen BEFORE sending to queue or race condition can occur
+		incrementWorkingCounter(component)
+		log.Debugf("adding subcomponent '%s' to queue with physical path '%s' and logical path '%s'\n", component.Name, component.PhysicalPath, component.LogicalPath)
+		queue <- component
+	}
+
 	// Manually enqueue the first root component
 	go func() {
-		component := prepareComponent(Component{
+		enqueue((prepareComponent(Component{
 			PhysicalPath: startingPath,
 			LogicalPath:  "./",
 			Config:       NewComponentConfig(startingPath),
-		})
-		enqueueComponent(component)
+		})))
 	}()
 
 	// Worker thread to pull from queue and call the iterator
 	go func() {
 		for component := range queue {
 			go func(component Component) {
-				completedSubcomponents := make(chan Component, len(component.Subcomponents)) // To keep track of which subcomponents have been enqueued; when len() == number of subcomponents, the parent is considered "complete"
-
 				// Call the iterator
 				results <- WalkResult{Error: iterator(component.PhysicalPath, &component)}
-
-				// A parent node is only completed after it's children have been enqueued.
-				go func() {
-					completedComponents := make([]Component, 0)
-					decrementIfNecessary := func() {
-						if len(completedComponents) == len(component.Subcomponents) {
-							markComponentAsVisited(component)
-						}
-					}
-					decrementIfNecessary() // Need to call once before loop in case component has no subcomponents
-					for subcomponent := range completedSubcomponents {
-						completedComponents = append(completedComponents, subcomponent)
-						decrementIfNecessary()
-					}
-				}()
 
 				// Range over subcomponents; preparing and enqueuing
 				for _, subcomponent := range component.Subcomponents {
@@ -320,9 +310,11 @@ func WalkComponentTree(startingPath string, environments []string, iterator Comp
 					}
 
 					log.Debugf("adding subcomponent '%s' to queue with physical path '%s' and logical path '%s'\n", subcomponent.Name, subcomponent.PhysicalPath, subcomponent.LogicalPath)
-					enqueueComponent(subcomponent)
-					completedSubcomponents <- subcomponent
+					enqueue(subcomponent)
 				}
+
+				// Decrement working counter; Must happen AFTER the subcomponents are enqueued
+				decrementWorkingCounter(component)
 			}(component)
 		}
 	}()
@@ -332,8 +324,8 @@ func WalkComponentTree(startingPath string, environments []string, iterator Comp
 
 // SynchronizeWalkResult will synchronize a channel of WalkResult to a list of visited Components.
 // It will return on the first Error encountered; returning the visited Components up until then and the error
-func SynchronizeWalkResult(results chan WalkResult) (components []Component, err error) {
-	components = make([]Component, 0)
+func SynchronizeWalkResult(results <-chan WalkResult) (components []Component, err error) {
+	components = []Component{}
 	for result := range results {
 		if result.Error != nil {
 			return components, err

